Spanner/src/StructType.php

```diff
@@ -0,0 +1,220 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Spanner;
+
+/**
+ * Represents a Struct Query Parameter type declaration.
+ *
+ * Example:
+ * ```
+ * use Google\Cloud\Spanner\StructType;
+ * use Google\Cloud\Spanner\Database;
+ * use Google\Cloud\Spanner\SpannerClient;
+ *
+ * $spanner = new SpannerClient();
+ * $database = $spanner->connect('my-instance', 'my-database');
+ *
+ * $res = $database->execute('SELECT @userStruct.firstName, @userStruct.lastName', [
+ *     'parameters' => [
+ *         'userStruct' => [
+ *             'firstName' => 'John',
+ *             'lastName' => 'Testuser'
+ *         ]
+ *     ],
+ *     'types' => [
+ *         'userStruct' => (new StructType())
+ *             ->add('firstName', Database::TYPE_STRING)
+ *             ->add('lastName', Database::TYPE_STRING)
+ *     ]
+ * ])->rows()->current();
+ *
+ * $fullName = $res['firstName'] . ' ' . $res['lastName']; // `John Testuser`
+ * ```
+ */
+class StructType
+{
+    /**
+     * @var array
+     */
+    private $fields = [];
+
+    /**
+     * @param array[] $fields An array containing a field definition. Each field
+     *        must be of form `[(string|null) $name, (int) $type, ArrayType|StructType|null $child]`.
+     */
+    public function __construct(array $fields = [])
+    {
+        foreach ($fields as $field) {
+            $field += [
+                'name' => null,
+                'type' => null,
+                'child' => null
+            ];
+
+            $this->add($field['name'], $field['type'], $field['child']);
+        }
+    }
+
+    /**
+     * Add a struct field definition.
+     *
+     * Unnamed struct fields may be defined by providing `null` as the first
+     * argument value, however you may find it more convenient to use the provided
+     * {@see Google\Cloud\Spanner\StructType::addUnnamed()} method.
+     *
+     * Example:
+     * ```
+     * $structType->add('firstName', Database::TYPE_STRING);
+     * ```
+     *
+     * ```
+     * // If a field is a struct or array, it must be explicitly defined.
+     * use Google\Cloud\Spanner\ArrayType;
+     * use Google\Cloud\Spanner\Database;
+     * use Google\Cloud\Spanner\StructType;
+     *
+     * // Create a struct which will be nested later.
+     * $customer = (new StructType)
+     *     ->add('name', Database::TYPE_STRING)
+     *     ->add('phone', Database::TYPE_STRING)
+     *     ->add('email', Database::TYPE_STRING)
+     *     ->add('lastOrderDate', Database::TYPE_DATE);
+     *
+     * // Create an array to nest within the customer type definition.
+     * $orderIds = new ArrayType(Database::TYPE_INT64);
+     * $customer->add('orderIds', Database::TYPE_ARRAY, $orderIds);
```

Comment:
passing Database::TYPE_ARRAY is redundant here. Is there a way to avoid that?

Reply:
I'd opted to preserve the typing of the 2nd argument for clarity's sake and for the sake of making intent explicit. I am open to this suggestion though if you feel it would be preferable.

Reply:
I'll change this to accept the ArrayType or StructType as the second argument.